### PR TITLE
[Hive]GraphSAGE

### PR DESCRIPTION
**Pull request by wangxiaoyunanne**
Work in progress for GraphSAGE doc

### CODE REVIEW

File: source/hive_graphSage.html.md [hunk 1]
```diff
@@ -10,68 +10,88 @@ search: true
 full_length: true
 ---
 
-<aside class="notice">
-  JDO notes, delete these when you copy this to `hive_yourworkflowname`: The goal of this report is to be useful to DARPA and to your colleagues. This is not a research paper. Be very honest. If there are limitations, spell them out. If something is broken or works poorly, say so. Above all else, make sure that the instructions to replicate the results you report are good instructions, and the process to replicate are as simple as possible; we want anyone to be able to replicate these results in a straightforward way.
-</aside>
+# GraphSAGE
 
-# GraphSage
+SAGE is a way to fit graphs into a neural networks: instead of getting the
```
Comment (jowens) [inline source/hive_graphSage.html.md]: I'd like a sentence on 'embedding'. What is it? This summary is a little too advanced for the non-expert here.

File: source/hive_graphSage.html.md [hunk 2]
```diff
@@ -10,68 +10,88 @@ search: true
...
-In GraphSage paper, figure 1 (2) shows aggregating feature information from neighbors, where read node is what we called `source` here and blue nodes are `child`, and green nodes are `leaf`. 
-
-The CPU psudocode is as follows:
+Gunrock's implementation for the 1st year is only the embedding / inferencing
+phase, without the training phase. It is based on Algorithm 2 with K=2 of GraphSage
+paper ("Inductive Representation Learning on Large Graphs",
+https://arxiv.org/abs/1706.02216).  
+
+Given a graph G, the inputs are per-vertex features, weight matrices W^k, and a
+non-linear activation function (ReLu); the output from SAGE is the embedding vector
```
Comment (jowens) [inline source/hive_graphSage.html.md]: Define "embedding vector".

File: source/hive_graphSage.html.md [hunk 3]
```diff
@@ -10,68 +10,88 @@ search: true
...
+of each vertex. The Gunrock implementation uses Mean aggregator at the moment;
```
Comment (jowens) [inline source/hive_graphSage.html.md]: What is an aggregator?

File: source/hive_graphSage.html.md [hunk 4]
```diff
@@ -10,68 +10,88 @@ search: true
...
+using other aggregators is possible with small modifications. Similarly, neighbors
+are selected randomly in the current implementation, and simple changes to the
+code can enable other selection methods, such as weighted uniform, importance
+sampling (used by FaseGCN) and random walk probability like DeepWalk or Node2Vec
+(used by PinSage).
```
Comment (jowens) [inline source/hive_graphSage.html.md]: We have a random-walk implementation, yes? Might be worth noting what it supports.

File: source/hive_graphSage.html.md [hunk 5]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
 in `tests/sage/bin`.
 
 The datasets are assumed to have been placed in `/raid/data/hive`, and converted
-to proper matrix market format (.mtx). At the time of testing,
-  `pokec` is available in that directory. 
+to proper matrix market format (.mtx). At the time of testing, `pokec`, `amazon`,
+'flickr', 'twitter' and 'cit-Patents' are available in that directory.
```
Comment (jowens) [inline source/hive_graphSage.html.md]: backticks on these datasets please.

File: source/hive_graphSage.html.md [hunk 6]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
+the DGX-1 machine, and using CUDA 10.0 with NVIDIA driver 410.66 on a Tesla V100
+GPU on Bowser.
 
-The testing is done with Gunrock using `dev-refactor` branch at commit `2699252`
-(Oct. 18, 2018), using CUDA 10.2 with NVIDIA driver 390.30.
 ### Running the application
 
+#### Application specific parameters
+```
+--Wa1 : std::string, default =
```
Comment (jowens) [inline source/hive_graphSage.html.md]: Helpful to say what the file format is here. I'm sure it's simple, but it should be specified.

File: source/hive_graphSage.html.md [hunk 7]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
+itself three times in the form of `Normalize(C x Wf + Mean(D) x Wa)`. Because of
+this simplicity, it is still possible to verify the implementation by visually
+inspecting the code. The resulted embeddings are also checked for the L2 norm,
+which should be close to 1 for every vertex.
 
 ## Performance and Analysis
 
-How do you measure performance? What are the relevant metrics? Runtime? Throughput? Some sort of accuracy/quality metric?
+The OpenMP and GPU implementations are tested for running time. It would be very
+useful to validate the successful rate of the whole pipeline with the training
```
Comment (jowens) [inline source/hive_graphSage.html.md]: Is there any validation at all for the parts we implemented? I guess the previous paragraph answers this ... just be very specific about what checks are done on the output and what checks are done against any reference code we have available to us.

File: source/hive_graphSage.html.md [hunk 8]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
+| twitter |  64 |  10 | 16384 |  7594.364 | 103753.961 | 13.66|
+| twitter |  64 |  25 | 16384 | 16212.790 | 162819.531 | 10.04|
+| twitter |  64 | 100 |  4096 | 73488.745 | 636224.625 |  8.66|
+| twitter | 128 |  10 | 16384 | 12753.125 | 147705.375 | 11.58|
+| twitter | 128 |  25 | 16384 | 28827.354 | 282497.438 |  9.80|
+| twitter | 128 | 100 |  4096 |134027.966 |1105741.500 |  8.25|
+
+The running time on Bowser with a Tesla V100 GPU
+
+| Dataset | F   | C   | B     | Gunrock GPU | OpenMP   | Speedup vs. CPU| Speed vs. P100 |
```
Comment (jowens) [inline source/hive_graphSage.html.md]: Speedup vs. P100   (not "speed")

those are impressive V100 speedups!

File: source/hive_graphSage.html.md [hunk 9]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
+| twitter | 128 |  25 |  8192 |  9744.686 | 267841.563 | 27.49| 2.93 |
+| twitter | 128 | 100 |  2048 | 45604.444 |1098039.375 | 24.08| 2.93 |
 
 ### Implementation limitations
 
-e.g.:
-
-- Size of dataset that fits into GPU memory (what is the specific limitation?)
-- Restrictions on the type/nature of the dataset
+* Memory usage * Gunrock's GPU implementation uses `B x (C x (Wf2.x + F + 1) +
```
Comment (jowens) [inline source/hive_graphSage.html.md]: I think it needs to be *Memory Usage* not * Memory Usage * or it won't render right (put the stars right next to the text)

File: source/hive_graphSage.html.md [hunk 10]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
 
 ### Gunrock implications
 
-What did we learn about Gunrock? What is hard to use, or slow? What potential Gunrock features would have been helpful in implementing this workflow?
+One thing that Gunrock does not provide, or intensionally hides, is the block
+level parallelism. However, it comes as handy when implementing the custom
+kernels for SAGE: each block can process a one dimension vector, with each
+thread holds one element, then use block level reduce to get the sum of those
+elements; this way is highly efficient, and actually reduces the parallelism,
+thus eventually the working set of data.
```
Comment (jowens) [inline source/hive_graphSage.html.md]: A couple of sentences on how this could be exposed in the Gunrock programming model would be useful. And the answer may be "it can't".

File: source/hive_graphSage.html.md [hunk 11]
```diff
@@ -10,68 +10,88 @@ search: true
...
-
-Instead of using B2, B1 and B0 name the nodes, we use `source`, `child`, and `leaf` to name three different groups of nodes. 
-In GraphSage paper, figure 1 (2) shows aggregating feature information from neighbors, where read node is what we called `source` here and blue nodes are `child`, and green nodes are `leaf`. 
-
-The CPU psudocode is as follows:
+Gunrock's implementation for the 1st year is only the embedding / inferencing
+phase, without the training phase. It is based on Algorithm 2 with K=2 of GraphSage
+paper ("Inductive Representation Learning on Large Graphs",
+https://arxiv.org/abs/1706.02216).  
+
```
Comment (jowens) [inline source/hive_graphSage.html.md]: So *I* need to understand why we only did embedding/inferencing. Can you justify this here? That was ... what the reference code did? I am out of the loop as far as why we picked this subset of things. My bad.

File: source/hive_graphSage.html.md [hunk 12]
```diff
@@ -10,68 +10,88 @@ search: true
...
+Given a graph G, the inputs are per-vertex features, weight matrices W^k, and a
+non-linear activation function (ReLu); the output from SAGE is the embedding vector
+of each vertex. The Gunrock implementation uses Mean aggregator at the moment;
+using other aggregators is possible with small modifications. Similarly, neighbors
+are selected randomly in the current implementation, and simple changes to the
+code can enable other selection methods, such as weighted uniform, importance
+sampling (used by FaseGCN) and random walk probability like DeepWalk or Node2Vec
+(used by PinSage).
+
+The pseudocode of Gunrock's implementation is as following. The GraphSage uses
```
Comment (jowens) [inline source/hive_graphSage.html.md]: It's unclear to me from this pseudocode what parts were Gunrock primitives and what parts were outside of Gunrock. Can you make that clearer?

File: source/hive_graphSage.html.md [hunk 13]
```diff
@@ -98,86 +118,397 @@ At this point, there should be an executable `test_sage_<CUDA version>_x86_64`
...
+larger than the combined GPU memory, it's possible to accumulate the features
+of leafs and children on CPU, transfer the data on to GPU, and perform the
+computation. Running time will increase significantly as compared to the cases when
+the full data can fit in the GPU memory; but whether that will increase above
+the OpenMP implementation is still unknown.
 
 ### Notes on other pieces of this workload
 
-Briefly: What are the important other (non-graph) pieces of this workload? Any thoughts on how we might implement them / what existing approaches/libraries might implement them?
+The main part of SAGE is actually the training process. How to connect the
```
Comment (jowens) [inline source/hive_graphSage.html.md]: Is training something straightforward within Gunrock? Outside of Gunrock? What do you mean by "connect"? This is an important section.

Comment (jowens) [reply source/hive_graphSage.html.md]: Please also add a "what is interesting from a research perspective" subsubsection.

File: source/hive_graphSage.html.md (hunk 1 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: Embedding means giving each node a vector representation.

File: source/hive_graphSage.html.md (hunk 2 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: according for algorithm, embedding vector is h_v

File: source/hive_graphSage.html.md (hunk 3 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: aggregator is the aggregator in algorithm, we use Mean() function in our code.

File: source/hive_graphSage.html.md (hunk 4 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: Node2vec is not used for inferencing. In training process it is needed.

File: source/hive_graphSage.html.md (hunk 1 above)
Comment (jowens) [reply source/hive_graphSage.html.md]: Put that in the doc! What is important about a vector representation? Why do we care? That needs to be front and center here.

File: source/hive_graphSage.html.md (hunk 2 above)
Comment (jowens) [reply source/hive_graphSage.html.md]: What IS an embedding vector though? (This ties back to the previous comment.) Give the reader some context about the _meaning_ of the output.

File: source/hive_graphSage.html.md (hunk 3 above)
Comment (jowens) [reply source/hive_graphSage.html.md]: But what _is_ an aggregator? Why do we have one? What are we aggregating? That is not in the description here.

File: source/hive_graphSage.html.md (hunk 7 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: I wrote a weak check function. 
For the forward process (inferencing), it will be very hard to have mistake.

File: source/hive_graphSage.html.md [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+---
+
+# GraphSAGE
+
+SAGE is a way to fit graphs into a neural networks: instead of getting the
+embedding of a vertex from all its neighbors' features as in conventional
+implementations, SAGE select some 1-hop neighbors, some 2-hop neighbors
+connected to those 1-hop neighbors, and compute the embedding based on the
+features of the 1-hop and 2-hop neighbors.
+
```
Comment (bkj) [inline source/hive_graphSage.html.md]: I think the name is "GraphSAGE", rather than just "SAGE", though I may be wrong.

File: source/hive_graphSage.html.md (hunk 11 above)
Comment (bkj) [reply source/hive_graphSage.html.md]: We decided to only do the inference step because the training step would require either 
 a) manually writing the backprop algorithm
 b) interfacing w/ an autodifferentiation library (eg Tensorflow/Pytorch/...)
as well as implementing all of the associated neural network dressing (batch loaders, optimizers, etc)

File: source/hive_graphSage.html.md (hunk 2 above)
Comment (bkj) [reply source/hive_graphSage.html.md]: Some motivation for why we'd be trying to train this NN would be good as well.  The example I always like to use is for document classification where you have network information about the authors (nodes are papers, links exist between nodes if the papers share an author).  You could use a model trained just on the text of the documents to predict it's topic.  Or you could just try to do community detection on the network.  But it turns out that graph NN models like GraphSAGE can take both of those kinds of features into account and do substantially better than either alone.

File: source/hive_graphSage.html.md (hunk 4 above)
Comment (bkj) [reply source/hive_graphSage.html.md]: Random walk isn't really used by GraphSAGE -- GraphSAGE does a 2-level hierarchical sampling of neighbors.  That has pieces that kindof smell like parts of a random walk, but the mechanics are different.

What is FaseGCN? (Just curious, I hadn't seen that one before).

File: source/hive_graphSage.html.md [hunk 15]
```diff
@@ -0,0 +1,514 @@
...
+(1.58.0 or 1.59.0) and Metis are installed; if not, changes need to be made in
+the Makefiles. DARPA's DGX-1 has both installed when the tests are performed.
+
+```
+git clone --recursive https://github.com/gunrock/gunrock/
+cd gunrock
+git checkout dev-refactor
+git submodule init
+git submodule update
+mkdir build
```
Comment (bkj) [inline source/hive_graphSage.html.md]: you can make this more concise w/ 
```bash
git clone --recursive https://github.com/gunrock/gunrock/ \
  -b dev-refactor
cd gunrock
mkdir build
cd build
... etc ...
```

File: source/hive_graphSage.html.md [hunk 16]
```diff
@@ -0,0 +1,514 @@
...
+Note that GraphSage is an inductive representation learning algorithm,
+so it reasonable to assume that there is no dangling vertices in the graph.
+*Before running GraphSAGE, please remove the dangling vertices from the graph. *
+In the case when dangling vertices are present, the dangling vertices themselves
+will be treated as their neighbors.  
+
+The testing is done with Gunrock using `dev-refactor` branch at commit `0ed72d5`
+(Oct. 25, 2018), using CUDA 9.2 with NVIDIA driver 390.30 on a Tesla P100 GPU in
+the DGX-1 machine, and using CUDA 10.0 with NVIDIA driver 410.66 on a Tesla V100
+GPU on Bowser.
```
Comment (bkj) [inline source/hive_graphSage.html.md]: What's Bowser?  One of your Davis machines?

File: source/hive_graphSage.html.md (hunk 6 above)
Comment (bkj) [reply source/hive_graphSage.html.md]: Are these datasets available somewhere in the proper format?  I can host them on the HIVE data server `hiveprogram.com/data/_gunrock/graphsage` or something, so that they're easily available to people that want to run the app.

File: source/hive_graphSage.html.md [hunk 17]
```diff
@@ -0,0 +1,514 @@
...
+| cit-Patents | 3774768 | 33037894 |
+| twitter | 7199978 | 43483326 |
+
+The running times in milliseconds are list below, for both machines. F stands
+for the length of features, C stands for the number of children per source,
+the same as the number of leafs per child, and B notes the batch size that
+produces the shortest running time on GPU among {512, 1024, 2048, 4096, 8192,
+16384}.
+
+The running time on DGX-1 with Tesla P100 GPU:
```
Comment (bkj) [inline source/hive_graphSage.html.md]: Are there comparisons against the HIVE reference implementation (runs on GPU via Pytorch) or the GT-USC implementation (OpenMP I think)?

https://gitlab.hiveprogram.com/ggillary/graphSAGE
https://gitlab.hiveprogram.com/gtusc/GraphSage

Comparison against the Pytorch implementation in particular is meaningful, because people typically use the GPU for this kind of thing anyway, but using much higher level APIs.

File: source/hive_graphSage.html.md [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+It's more interesting when comparing the running time on V100 and P100: V100 is
+about 3X faster than P100 when running SAGE. This large performance difference
+is caused by the different limiting factors when running SAGE on these two GPUs;
+details are in the the Performance limitations section. Compared to OpenMP, V100
+is about 20X to 30X faster.
+
+### Performance limitations
+
+Profiling SAGE with the pocket dataset on a Titan Xp GPU (profiling on P100
+caused internal error in the profiler itself; Titan Xp has roughly the same SM
```
Comment (bkj) [inline source/hive_graphSage.html.md]: pokec, I think

File: source/hive_graphSage.html.md [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+into child-centric and source-centric parts, and exchange data in between. It
+should be scalable, and easy to implement.
+
+### Notes on dynamic graphs
+
+SAGE does not have a dynamic graph component, but it should able to work on a
+dynamic graph. Some of the data may be reusable if the graph has not been
+significantly changed; but the resulted memory requirement to store the intermedia
+data may make the data reuse impossible.
+
```
Comment (bkj) [inline source/hive_graphSage.html.md]: One of the supposed advantages of GraphSAGE over other graph neural networks is that it's inductive rather than transductive -- you can add new nodes to the graph and run the same inference procedure.  So at least there's wouldn't be trouble extending the GraphSAGE _algorithm_ to dynamic graphs.

File: source/hive_graphSage.html.md (hunk 13 above)
Comment (wangxiaoyunanne) [reply source/hive_graphSage.html.md]: The inferencing process is the forward  propagation process, the training process is the backward propagation. So for training we need a loss function and an optimizer (say SGD), use then use them to update parameters.  Training should be outside of Gunrock. 
From research perspective, accelerating training process is more interesting. Because when the graphs are large, there is a random access bottleneck.

**Comment by sgpyc** [comment]
Thanks a lot to @jowens and @bkj 's comments. They are addressed in the new commits
